### ISSUE TITLE
Avoiding explicit call to DATAPTR in vector constructors

### ISSUE THREAD

**Issue by jimhester** [RcppCore/Rcpp#955]
Rcpp calls `DATAPTR()` on all vectors passed to the Rcpp integer and numeric vector classes.

https://github.com/RcppCore/Rcpp/blob/d55f9f1e206e7c0452fe4f6f97c840109eaab26e/inst/include/Rcpp/internal/r_vector.h#L32

which means if these vectors are Altrep vectors the full values are materialized when the object is constructed. This is unfortunate, as it means if any package is using Rcpp they effectively lose the benefits of Altrep. This can be seen in the following example using the compact integer sequences in R 3.5+ (notice how the altrep is shown as compact in the first inspect and then expanded in the second).

```c++
// altrep.cpp
#include <Rcpp.h>

using namespace Rcpp;

/* Just return the SEXP unchanged */
//[[Rcpp::export]]
SEXP convertSEXP(SEXP data) {
  return data;
}

/* Pass the SEXP to a Rcpp vector then return that */
//[[Rcpp::export]]
SEXP convertRcpp(SEXP data) {
  switch(TYPEOF(data)) {
    case STRSXP:
      return CharacterVector(data);
    case INTSXP:
      return IntegerVector(data);
    case REALSXP:
      return NumericVector(data);
  }
  stop("wrong type");
}
```
``` r
Rcpp::sourceCpp("~/p/vroom/altrep.cpp")
x <- 1:100
invisible(convertSEXP(x))
.Internal(inspect(x))
#> @7fd5a90ddd20 13 INTSXP g0c0 [NAM(3)]  1 : 100 (compact)
invisible(convertRcpp(x))
.Internal(inspect(x))
#> @7fd5a90ddd20 13 INTSXP g0c0 [NAM(3)]  1 : 100 (expanded)
```

My [vroom](https://github.com/jimhester/vroom) package uses altrep to read files lazily and exhibits the same behavior. (Note the thing to look for in this example is `materialized = T` and also note that CharacterVector objects do not have the same issue).

<details>
<summary> vroom example </summary>

``` r
Rcpp::sourceCpp("~/p/vroom/altrep.cpp")

# Generate a 10x5 table with columns of type, character, integer, double,
# POSIXct and Date, then write it to a file
set.seed(42)
types <- "cidTD"
x <- vroom::gen_tbl(10, 5, col_types = types)
readr::write_tsv(x, "altrep.tsv")

# Read it with the vroom package
y <- vroom::vroom("altrep.tsv", col_types = types)

# Look at the results, the columns are read as un-materialized altrep vectors
.Internal(inspect(y, 1, 6))
#> @7fc043c218e8 19 VECSXP g0c4 [OBJ,NAM(3),ATT] (len=5, tl=0)
#>   @7fc03e667b08 16 STRSXP g0c0 [NAM(3)] vroom_chr (len=10, materialized=F)
#>   @7fc03e6678a0 13 INTSXP g0c0 [NAM(3)] vroom_int (len=10, materialized=F)
#>   @7fc03e667638 14 REALSXP g0c0 [NAM(3)] vroom_dbl (len=10, materialized=F)
#>   @7fc03e6673d0 14 REALSXP g0c0 [OBJ,NAM(3),ATT] vroom_dttm (len=10, materialized=F)
#>   ATTRIB:
#>     @7fc03e667328 02 LISTSXP g0c0 [] 
#>   @7fc03e667088 14 REALSXP g0c0 [OBJ,NAM(3),ATT] vroom_date (len=10, materialized=F)
#>   ATTRIB:
#>     @7fc03e666fe0 02 LISTSXP g0c0 [] 
#> ATTRIB:
#>   @7fc03d2c2250 02 LISTSXP g0c0 [] 
#>     TAG: @7fc03c81eb00 01 SYMSXP g1c0 [MARK,LCK,gp=0x4000] "names" (has value)
#>     @7fc043c21f78 16 STRSXP g0c4 [NAM(3)] (len=5, tl=0)
#>     TAG: @7fc03c81e8d0 01 SYMSXP g1c0 [MARK,NAM(3),LCK,gp=0x4000] "row.names" (has value)
#>     @7fc03d6ad978 13 INTSXP g0c1 [] (len=2, tl=0) -2147483648,-10
#>     TAG: @7fc03c81efd0 01 SYMSXP g1c0 [MARK,NAM(3),LCK,gp=0x6000] "class" (has value)
#>     @7fc03d099d88 16 STRSXP g0c3 [NAM(1)] (len=3, tl=0)

# If using the C interface just roundtripping them does not materialize the
# vectors (as DATAPTR()) is never called.
for (i in seq_along(y)) convertSEXP(y[[i]])
.Internal(inspect(y, 1, 6))
#> @7fc043c218e8 19 VECSXP g0c4 [OBJ,MARK,NAM(3),ATT] (len=5, tl=0)
#>   @7fc03e667b08 16 STRSXP g0c0 [MARK,NAM(3)] vroom_chr (len=10, materialized=F)
#>   @7fc03e6678a0 13 INTSXP g0c0 [MARK,NAM(3)] vroom_int (len=10, materialized=F)
#>   @7fc03e667638 14 REALSXP g0c0 [MARK,NAM(3)] vroom_dbl (len=10, materialized=F)
#>   @7fc03e6673d0 14 REALSXP g0c0 [OBJ,MARK,NAM(3),ATT] vroom_dttm (len=10, materialized=F)
#>   ATTRIB:
#>     @7fc03e667328 02 LISTSXP g0c0 [MARK] 
#>   @7fc03e667088 14 REALSXP g0c0 [OBJ,MARK,NAM(3),ATT] vroom_date (len=10, materialized=F)
#>   ATTRIB:
#>     @7fc03e666fe0 02 LISTSXP g0c0 [MARK] 
#> ATTRIB:
#>   @7fc03d2c2250 02 LISTSXP g0c0 [MARK] 
#>     TAG: @7fc03c81eb00 01 SYMSXP g1c0 [MARK,LCK,gp=0x4000] "names" (has value)
#>     @7fc043c21f78 16 STRSXP g0c4 [MARK,NAM(3)] (len=5, tl=0)
#>     TAG: @7fc03c81e8d0 01 SYMSXP g1c0 [MARK,NAM(3),LCK,gp=0x4000] "row.names" (has value)
#>     @7fc03d6ad978 13 INTSXP g0c1 [MARK] (len=2, tl=0) -2147483648,-10
#>     TAG: @7fc03c81efd0 01 SYMSXP g1c0 [MARK,NAM(3),LCK,gp=0x6000] "class" (has value)
#>     @7fc03d099d88 16 STRSXP g0c3 [MARK,NAM(3)] (len=3, tl=0)

# But as soon as you use a Rcpp vector class the numeric columns are materialized
for (i in seq_along(y)) convertRcpp(y[[i]])
.Internal(inspect(y, 1, 6))
#> @7fc043c218e8 19 VECSXP g0c4 [OBJ,MARK,NAM(3),ATT] (len=5, tl=0)
#>   @7fc03e667b08 16 STRSXP g0c0 [MARK,NAM(3)] vroom_chr (len=10, materialized=F)
#>   @7fc03e6678a0 13 INTSXP g0c0 [MARK,NAM(3)] vroom_int (len=10, materialized=T)
#>   @7fc03e667638 14 REALSXP g0c0 [MARK,NAM(3)] vroom_dbl (len=10, materialized=T)
#>   @7fc03e6673d0 14 REALSXP g0c0 [OBJ,MARK,NAM(3),ATT] vroom_dttm (len=10, materialized=T)
#>   ATTRIB:
#>     @7fc03e667328 02 LISTSXP g0c0 [MARK] 
#>   @7fc03e667088 14 REALSXP g0c0 [OBJ,MARK,NAM(3),ATT] vroom_date (len=10, materialized=T)
#>   ATTRIB:
#>     @7fc03e666fe0 02 LISTSXP g0c0 [MARK] 
#> ATTRIB:
#>   @7fc03d2c2250 02 LISTSXP g0c0 [MARK] 
#>     TAG: @7fc03c81eb00 01 SYMSXP g1c0 [MARK,LCK,gp=0x4000] "names" (has value)
#>     @7fc043c21f78 16 STRSXP g0c4 [MARK,NAM(3)] (len=5, tl=0)
#>     TAG: @7fc03c81e8d0 01 SYMSXP g1c0 [MARK,NAM(3),LCK,gp=0x4000] "row.names" (has value)
#>     @7fc03d6ad978 13 INTSXP g0c1 [MARK] (len=2, tl=0) -2147483648,-10
#>     TAG: @7fc03c81efd0 01 SYMSXP g1c0 [MARK,NAM(3),LCK,gp=0x6000] "class" (has value)
#>     @7fc03d099d88 16 STRSXP g0c3 [MARK,NAM(3)] (len=3, tl=0)
```
</details>

Ideally Rcpp would avoid using `DATAPTR()` and instead use the appropriate Altrep aware accessor functions, e.g. `REAL_ELT` / `INTEGER_ELT` where applicable.

I would be willing to work on this if there was interest.

**Comment by eddelbuettel** [comment]
There is certainly interest!  We contemplated hanging out a GSoC shingle for this but weren't sure we'd get a student. I am fine with incremental fixes too -- correcting _some_ case beats not changing any.

**Comment by eddelbuettel** [comment]
But there are other concerns too meaning that there are some things we cannot fix easily due to the way R does things right now (even with ALTREP) and how Rcpp does things -- see the long discussion in #943.

**Comment by jimhester** [comment]
So at least the refcount bit which always increments  of #943 seems due to Rcpp's use of a R closure to do the argument matching. If instead Rcpp generated a C shim function instead that handled the argument matching that seems like a way around the problem.

**Comment by eddelbuettel** [comment]
Yes, I thought about that area but didn't immediately think of a fix.  Maybe you have a better vision here -- that is definitely scope for improvement too.  "One shim to rule them all" would be nicer but I presume it would be one shim each?

**Comment by eddelbuettel** [comment]
@jimhester as it has been a month (and as you took care of things in `vroom` directly) is still "open" for you, or is it more "dormant" ?

**Comment by jimhester** [comment]
I took care of things in vroom by only using altrep for character vectors, which don't exhibit this behavior, so really that was sidestepping the issue.

I did some [benchmarking](https://rpubs.com/jimhester/altrep) of the overhead in using `REAL_ELT()` vs using `REAL()` to get the pointer and also with a block altrep interface.

However the Rcpp codebase around the vector iterators is fairly sprawling throughout the Rcpp codeback and I wasn't entirely clear where the best place to make the changes would be.

So for now I have shelved it in interest of time, but I will likely come back to it in the future, and I do still think it is something that needs to be addressed in some fashion going forward.
